Deboostify Crow

luca-schlecker opened the pull request:
This Pr removes `Boost` as a dependency and will instead rely on the standalone `asio` library.
CI now installs `libasio-dev` instead of `libboost-all-dev` and the documentation resembles that change. (Although the MacOS section could use a double-check, I have no such device on hand)

### Proper testing necessary
Unittests are all passing but I think some real-life testing does no harm.

### Removed Boost functions and types and their replacements:
- `boost::lexical_cast` becomes `crow::utility::lexical_cast`. Its implementation relies on `std::stringstream` for converting between types.
- `boost::hash_combine` is replaced by a private function. Its implementation idea stems from around the internet and can be found in multiple sources.
- `boost::iequals` is replaced by `crow::utility::string_equals` taking two mandatory `std::string` arguments and one optional `bool` controlling the case-sensitivity.
- `boost::system::error_code` is replaced by `asio::error_code` which is an alias for `std::error_code`.
- Every `boost::asio::` function is replaced with their standalone `asio::` counterpart.
- `boost::array` is replaced with `std::array`.
- `boost::posix_time` is replaced with a `std::chrono` equivalent.
- `boost::*_comparable` is dropped and the respective operators are implemented manually.
- `boost::lexicographical_compare` is replaced with `std::lexicographical_compare`.
- `boost::optional` is replaced with `std::unique_ptr`.
- `boost::trim` is replaced with `crow::utility::trim`.

### Introduced functions (\*explicitly\* tested?):
- `private void crow::ci_hash::hash_combine`. (❌)
- `private bool crow::json::detail::r_string::equals`. (❌)
- Copy-constructor for `crow::CookieParser::Cookie`. (❌)
- `bool crow::utility::string_equals` (✅)
- `T crow::utility::lexical_cast` (✅)
- `std::string crow::utility::trim` (✅)

The-EDev commented on include/crow/json.h:
```diff
@@ -189,27 +187,57 @@ namespace crow
             }
 
             inline bool operator>(const r_string& l, const std::string& r)
             {
-                return boost::lexicographical_compare(r, l);
+                return std::lexicographical_compare(l.begin(), l.end(), r.begin(), r.end());
+            }
+
+            inline bool operator>(const std::string& l, const r_string& r)
+            {
+                return std::lexicographical_compare(l.begin(), l.end(), r.begin(), r.end());
             }
```
looks like we're missing operator > for both being `r_string`

luca-schlecker replied:
Seems like I missed that one. Will be added. 👍

The-EDev commented on include/crow/json.h:
```diff
@@ -212,17 +212,44 @@ namespace crow
             inline bool operator==(const r_string& l, const r_string& r)
             {
-                return std::equal(l.begin(), l.end(), r.begin(), r.end());
+                if (l.size() != r.size())
+                    return false;
+
+                for (size_t i = 0; i < l.size(); i++)
+                {
+                    if (*(l.begin() + i) != *(r.begin() + i))
+                        return false;
+                }
+
+                return true;
```
would a template utility function that uses `std::equals` for newer C++ and this function for C++11 work?

luca-schlecker replied:
I guess it would but if the algorithm has to be there anyway then why not use it.

The-EDev replied:
is there a way to have this algorithm in a way where it's not repeated for all 3 `==` operators?

luca-schlecker replied:
I could see a private templated function that is being called by these operators.

The-EDev commented on include/crow/http_server.h:
```diff
@@ -124,9 +124,9 @@ namespace crow
+                tick_timer_.expires_after(std::chrono::milliseconds(tick_interval_.count()));
                 tick_timer_.async_wait(
-                  [this](const boost::system::error_code& ec) {
+                  [this](const std::error_code& ec) {
```
line 113 in `task_timer.h` to be specific.

luca-schlecker replied:
Will change for consistency and to show intentions of the code. 👍

luca-schlecker commented on include/crow/utility.h:
```diff
@@ -780,5 +781,27 @@ namespace crow
+            if (l.length() != r.length())
+                return false;
+
+            for (size_t i = 0; i < l.length(); i++)
+            {
+                if (case_sensitive)
+                    equal &= (l[i] == r[i]);
+                else
+                    equal &= (std::toupper(l[i]) == std::toupper(r[i]));
+            }
```
Missed that case, thank you. 👍

luca-schlecker commented:
I'll have to look into that unittest again. Passing fine on my system.
Unfortunately, I had to make a change to `query_string.h` and in the process vscode auto-formatted everything according to Crow's guidelines. The `yml` files were also formatted on save, reversed that change though.

The-EDev reviewed (commented):
Aside from the comments, I couldn't find anything concerning from looking at the code. I'll do a few local tests (to see if it compiles with GCC 4.8 for example) and we'll see what to do from there :)

The-EDev commented on docs/getting_started/setup/linux.md:
```diff
@@ -3,7 +3,7 @@ Here's how you can install Crow on your favorite GNU/Linux distro.
 ### Requirements
  - C++ compiler with at least C++11 support.
- - boost library & development headers (1.64 or later).
+ - Asio development headers.
```
Might be worth including the minimum ASIO version (`1.10.9` was the version included in boost `1.64`)

luca-schlecker replied:
I guess a minimum version should be included. What function limits the version? Why was boost 1.64 chosen?

The-EDev replied:
boost 1.64 was chosen I believe because that's the version conan used.

I looked at the changelog and it seems like 1.10.8 was the first to support OpenSSL 1.1, so 1.10.9 is either the oldest or second oldest version that Crow can work with.

[source](https://www.boost.org/doc/libs/1_79_0/doc/html/boost_asio/history.html#boost_asio.history.asio_1_10_9___boost_1_64)

luca-schlecker replied:
Then I'll make sure to mention it. 👍

The-EDev commented:
```sh
$ g++-4.8 example_with_all.cpp -std=c++11 -lpthread -o examp
In file included from example_with_all.cpp:1:0:
crow_all.h: In member function ‘std::string crow::CookieParser::Cookie::dump() const’:
crow_all.h:5537:27: error: ‘put_time’ is not a member of ‘std’
                        << std::put_time(expires_at_.get(), HTTP_DATE_FORMAT);
                           ^
```

it looks like the code doesn't compile with GCC4.8's c++11, which is either a bug in GCC or a missing feature from c++11

luca-schlecker commented:
This PR doesn't add the `std::put_time` call though, it was there before.

The-EDev commented:
So here's what I did, I modified the `helloworld` example to run multithreaded and then again to run no logs. Then I ran `wrk` with 8 connections and 8 threads on both a `master` build and a boostless one. The results may not be accurate as to Crow's actual performance, but they are useful in measuring the difference between boostless and boostful Crow:
```
-- WITH LOG --

- NoBoost -
Running 1m test @ http://0.0.0.0:18080/
  8 threads and 8 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   183.56us  177.83us  16.45ms   95.69%
    Req/Sec     5.75k   188.66     6.88k    75.31%
  2750033 requests in 1.00m, 351.43MB read
Requests/sec:  45757.82
Transfer/sec:      5.85MB

- Master -
Running 1m test @ http://0.0.0.0:18080/
  8 threads and 8 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   655.24us  166.19us  12.31ms   80.37%
    Req/Sec     1.53k   276.82     8.51k    70.84%
  730762 requests in 1.00m, 93.39MB read
Requests/sec:  12159.17
Transfer/sec:      1.55MB


-- NO LOG --

- NoBoost -
Running 1m test @ http://0.0.0.0:18080/
  8 threads and 8 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    58.25us  276.31us  23.32ms   99.38%
    Req/Sec    21.20k   750.27    31.28k    87.37%
  10134223 requests in 1.00m, 1.26GB read
Requests/sec: 168623.70
Transfer/sec:     21.55MB

- Master -
Running 1m test @ http://0.0.0.0:18080/
  8 threads and 8 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   615.59us  262.44us  20.61ms   92.73%
    Req/Sec     1.64k   375.20     6.07k    68.39%
  783159 requests in 1.00m, 100.08MB read
Requests/sec:  13031.10
Transfer/sec:      1.67MB
```

These results show an improvement of 400% to 1000%. I couldn't believe the results so I ran the boost tests 3 more times, and got the same result.

if I'm not doing something horribly wrong (which I probably am) This PR is an extreme performance "boost" (Ha Ha) as well as an improvement in Crow's dependencies.

The-EDev commented:
The only thing left is the asio version, everything else seems fine :)

The-EDev reviewed (approved):
Everything seems to be in order, PR seems to have no effect on compiler version and it doesn't seem to be any slower than the non-boost version.

@mrozigor Would you like to take a look or share some feedback? It's a pretty large PR so I would very much appreciate your opinion on it.

mrozigor commented on include/crow/middlewares/cookie_parser.h:
```diff
@@ -93,14 +93,14 @@ namespace crow
             Cookie& expires(const std::tm& time)
             {
-                expires_at_ = time;
+                expires_at_ = std::unique_ptr<std::tm>(new std::tm(time));
```
Change it to `make_unique`?

luca-schlecker replied:
I believe `std::make_unique` wasn't added till `C++14`.

mrozigor replied:
So maybe it is the time to bump up at least to C++14? ;)

The-EDev replied:
@mrozigor is `make_unique` different from what we have here?

luca-schlecker replied:
Basically not, but it's nicer.

The-EDev replied:
Then I would be against deprecating C++11 support. I know it's old, but there's too much infrastructure still relying on it.

The-EDev replied:
Well github just deleted my comment...

C++14 breaks compatibility for minor changes, this is a problem because downstream projects may not be able to just simply upgrade to a new standard and possibly a new compiler too. We've already run into issues with GCC 4.8 and GCC 5 support (I believe we'll end up dropping 4.8 soon because of #463).

The work's already been done, and it seems to be a lot more leak-proof than using simple pointers. That being said if you're still not convinced @mrozigor  we can perhaps use a macro or function that uses this implementation for C++11 and `std::make_unique` for later versions.

mrozigor replied:
No, we can merge it as is. I'm just curious if is there anyone who is not able to update to newer gcc/clang? Even our IoT project was able to use GCC in version 9.

Anyway, it's just my thought about this, feel free to ignore it ;)

The-EDev replied:
Actually we had someone who wanted to downgrade GCC from 4.8 all the way to 4.4 [on gitter](https://gitter.im/crowfork/community?at=62a4cceedeea5616bbf1a292) 😅. I eventually had to turn him/her down because 4.4 didn't properly support c++11.

mrozigor replied:
TBH we shouldn't keep older C++ version just because someone is using very old software ;)

mrozigor commented:
> Everything seems to be in order, PR seems to have no effect on compiler version and it doesn't seem to be any slower than the non-boost version.
> 
> @mrozigor Would you like to take a look or share some feedback? It's a pretty large PR so I would very much appreciate your opinion on it.

I've look into code and it looks pretty good. And great work to remove unnecessary dependency! ;)

The-EDev commented:
I'll take one last look on the MacOS instructions and try to follow them to see if there are any problems.

luca-schlecker commented:
There is something wrong with the `crow::utility::trim` function I have overlooked.
I'll fix it.

luca-schlecker commented:
Fixed the `trim` usage and removed two redundant whitespace checks in `CookieParser`. Unit tests are still passing on my system.
This Pr should now be ready for merging. 🚀

The-EDev commented:
There's an issue compiling on macos relating to query_string, I'm looking into it

The-EDev reviewed (approved):
The bug is Xcode related, Compiling with g++ worked flawlessly (although I had to specify `-std=c++11`).

I apologize for this taking too long, merging now.
